File: tools/conformance/src/validators/typescript/validator.ts
import mustache from "mustache";
import { createSourceFile, ScriptKind, ScriptTarget } from "typescript";

import { validateDepthFirstSearch } from "./abstract-syntax-tree";

import type { ConformanceError, ConformanceErrorLanguage } from "./types";

function resolveScriptKind(filename: string): ScriptKind {
  const extension = filename.slice(filename.lastIndexOf("."));
  switch (extension) {
    case ".tsx": {
      return ScriptKind.TSX;
    }
    case ".ts": {
      return ScriptKind.TS;
    }
    case ".jsx": {
      return ScriptKind.JSX;
    }
    case ".mjs": {
      return ScriptKind.JS;
    }
    case ".cjs": {
      return ScriptKind.JS;
    }
    case ".js": {
      return ScriptKind.JS;
    }
    default: {
      return ScriptKind.TS;
    }
  }
}

function resolveLanguage(filename: string): ConformanceErrorLanguage {
  const extension = filename.slice(filename.lastIndexOf("."));
  switch (extension) {
    case ".ts":
    case ".tsx": {
      return "typescript";
    }
    case ".js":
    case ".jsx":
    case ".mjs":
    case ".cjs": {
      return "javascript";
    }
    default: {
      return "typescript";
    }
  }
}

/**
 * Validates that a generated TypeScript file is a structural superset of its
 * Mustache template by comparing their parsed ASTs node-by-node.
 *
 * The template is first rendered with data via Mustache, then both the rendered
 * template and instance are parsed into TypeScript ASTs. A depth-first walk
 * checks that every node in the template exists somewhere in the instance at
 * the same depth (superset, not equality — the instance may contain extra nodes
 * not present in the template). Type annotations, decorator arguments, import
 * specifiers, and named declarations are all checked; empty method bodies and
 * array literals are not (recursion stops where the template has no children).
 *
 * Comments are validated in template order via the TypeScript trivia API. TODO
 * comments match loosely (any `// ... TODO ...` line); all others must match
 * exactly.
 *
 * Use this function when `template` and `instance` are already in memory.
 */
export function validateTypescriptConformance(args: {
  data: Record<string, unknown>;
  filename: string;
  instance: string;
  template: string;
}): {
  errors: ConformanceError[];
} {
  const { instance, template, data, filename } = args;

  const scriptKind = resolveScriptKind(filename);
  const language = resolveLanguage(filename);
  const templateFile = createSourceFile(
    filename,
    mustache.render(template, data),
    ScriptTarget.Latest,
    true,
    scriptKind,
  );
  const instanceFile = createSourceFile(
    filename,
    instance,
    ScriptTarget.Latest,
    true,
    scriptKind,
  );

  const errors = validateDepthFirstSearch({
    templateNode: templateFile,
    instanceNode: instanceFile,
    instanceFile,
    language,
  });

  return { errors };
}

File: tools/conformance/src/modules/validator/python/validator.py
"""🐍 Python file conformance validator using abstract syntax trees."""

import ast

from python.abstract_syntax_tree import validate_depth_first_search
from python.comments import validate_comments
from python.template import render_template
from python.types import ConformanceError


def validate_python_conformance(*, data: dict, filename: str, instance: str, template: str) -> dict:
    """Validates that a generated Python file is a structural superset of its Mustache template."""
    rendered = render_template(template=template, data=data)
    try:
        template_tree = ast.parse(rendered, filename=filename)
    except SyntaxError as e:
        return {
            "errors": [
                ConformanceError(
                    error_type="code",
                    language="python",
                    message=f"Template syntax error: {e}",
                    fix="Fix the syntax error in the template.",
                )
            ]
        }
    try:
        instance_tree = ast.parse(instance, filename=filename)
    except SyntaxError as e:
        return {
            "errors": [
                ConformanceError(
                    error_type="code",
                    language="python",
                    message=f"Instance syntax error: {e}",
                    fix="Fix the syntax error in the instance file.",
                )
            ]
        }
    structural_errors = validate_depth_first_search(
        template_tree, instance_tree, rendered, instance, filename
    )
    comment_errors = validate_comments(rendered, instance)
    return {"errors": [*structural_errors, *comment_errors]}
